haystack/preview/components/reader.py

```diff
@@ -0,0 +1,151 @@
+
+
+@component
+class ExtractiveReader:
+    def __init__(
+        self, reader: Union[Path, str], device: Optional[str] = None, max_seq_length: int = 384, top_k: int = 10
+    ) -> None:
+        self.reader = reader
+        self.device = device
+        self.loaded = False
+        self.max_seq_length = max_seq_length
+        self.top_k = top_k
+
+    def warm_up(self):
+        if not self.loaded:
+            torch_and_transformers_import.check()
+            if torch.cuda.is_available():
+                self.device = "cuda:0"
+            else:
+                self.device = "cpu:0"
+            self.model = AutoModelForQuestionAnswering.from_pretrained(self.reader).to(self.device)
+            self.tokenizer = Tokenizer.from_pretrained(self.reader)
+
+    def _flatten(self, queries: List[str], documents: List[List[Document]]) -> Tuple[List[str], List[str]]:
+        flattened_queries = [query for documents_, query in zip(documents, queries) for _ in documents_]
+        flattened_documents = [document.content for documents_ in documents for document in documents_]
+        return flattened_queries, flattened_documents
+
+    def _preprocess(
+        self, queries: List[str], documents: List[str], max_seq_length: int
+    ) -> Tuple[torch.tensor, torch.tensor, torch.tensor, List[Encoding]]:
+        self.tokenizer.enable_padding(length=max_seq_length)
+        self.tokenizer.enable_truncation(max_seq_length)
+
+        encodings = self.tokenizer.encode_batch(list(zip(queries, documents)))
+
+        input_ids = torch.tensor([encoding.ids for encoding in encodings]).to(self.device)
+        attention_mask = torch.tensor([encoding.attention_mask for encoding in encodings]).to(self.device)
+        sequence_ids = torch.tensor(
+            [[id_ if id_ is not None else -1 for id_ in encoding.sequence_ids] for encoding in encodings]
+        ).to(self.device)
+
+        return input_ids, attention_mask, sequence_ids, encodings
+
+    def _postprocess(
+        self,
+        output,
+        sequence_ids: torch.tensor,
+        attention_mask: torch.tensor,
+        top_k: int,
+        encodings: List[Encoding],
+        max_seq_length: int,
+    ) -> Tuple[List[List[int]], List[List[int]], List[List[float]]]:
+        start = output.start_logits
+        end = output.end_logits
+        mask = sequence_ids == 1
+        mask[..., 0] = True
+        mask = torch.logical_and(mask, attention_mask == 1)
+        start = torch.where(mask, start, -torch.inf)
+        end = torch.where(mask, end, -torch.inf)
+        start = torch.softmax(start, -1)
+        end = torch.softmax(end, -1)
+        start = start.unsqueeze(-1)
+        end = end.unsqueeze(-2)
+
+        probabilities = start * end  # shape: (batch_size, seq_length (start), seq_length (end))
+        masked_probabilities = torch.triu(probabilities, diagonal=1)
+        masked_probabilities[..., 0, 0] = probabilities[..., 0, 0]
+        masked_probabilities[..., 0, 1:] = 0
```

Comment:
It's slightly dodgy to do this after applying softmax because then all possible answers will never add up to 1. The only solution I see is to repeat start and end and apply softmax separately. However, this would require more memory and FARMReader also didn't go that far.

Reply:
If this is the same approach FARMReades has taken, then I'd add a comment above this code explaining the issue and I'll leave it as it is :+1: